include/dali/core/tensor_layout.h

```diff
@@ -0,0 +1,490 @@
+struct LayoutInfo {
+  DALI_HOST_DEV
+  static bool HasSampleDim(const TensorLayout &tl) {
+    // sample dim may only appear at outermost level
+    return !tl.empty() && tl[0] == 'N';
+  }
+  DALI_HOST_DEV
+  static int DimIndex(const TensorLayout &tl, char dim_symbol) {
+    return tl.find(dim_symbol);
+  }
+};
+
+/** @brief Provides functions for querying TensorLayout properties,
+ *         assuming that the layout describes an image
+ */
+struct ImageLayoutInfo : LayoutInfo {
+  DALI_HOST_DEV
+  static int NumSpatialDims(const TensorLayout &tl) {
+    int s = 0;
+    for (int i = 0; i < tl.ndim(); i++) {
+      switch (tl[i]) {
+      case 'D':
+      case 'H':
+      case 'W':
+        s++;
+        break;
+      default:
+        break;
+      }
+    }
+    return s;
+  }
+  DALI_HOST_DEV
+  static int ChannelDimIndex(const TensorLayout &tl) {
+    return DimIndex(tl, 'C');
+  }
+  DALI_HOST_DEV
+  static bool HasChannel(const TensorLayout &tl) {
+    return ChannelDimIndex(tl) >= 0;
+  }
+  DALI_HOST_DEV
+  static bool Is2D(const TensorLayout &tl) {
+    return NumSpatialDims(tl) == 2;
+  }
+  DALI_HOST_DEV
+  static bool Is3D(const TensorLayout &tl) {
+    return NumSpatialDims(tl) == 3;
+  }
+  DALI_HOST_DEV
+  static bool IsChannelFirst(const TensorLayout &tl) {
+    return tl[0] == 'C' || (tl[0] == 'N' && tl[1] == 'C');
+  }
+  DALI_HOST_DEV
+  static bool IsChannelLast(const TensorLayout &tl) {
+    return !tl.empty() && tl[tl.ndim()-1] == 'C';
+  }
+  DALI_HOST_DEV
+  static bool IsImage(const TensorLayout &tl) {
+    return NumSpatialDims(tl) >= 2;
+  }
+};
+
+/** @brief Provides functions for querying TensorLayout properties,
+ *         assuming that the layout describes a video
+ */
+struct VideoLayoutInfo : ImageLayoutInfo {
+  /** @brief Returns the index of the dimension referring to frames */
+  DALI_HOST_DEV
+  static int FrameDimIndex(const TensorLayout &tl) {
+    return DimIndex(tl, 'F');
+  }
+  DALI_HOST_DEV
+  static bool IsChannelFirst(const TensorLayout &tl) {
+    return tl[FrameDimIndex(tl)+1] == 'C';
+  }
+  DALI_HOST_DEV
+  static bool IsSequence(const TensorLayout &tl) {
+    return tl.contains('F');
+  }
+  DALI_HOST_DEV
+  static bool IsVideo(const TensorLayout &tl) {
+    return IsSequence(tl) && IsImage(tl);
+  }
+  DALI_HOST_DEV
+  static bool IsStillImage(const TensorLayout &tl) {
+    return !IsSequence(tl) && IsImage(tl);
+  }
+  DALI_HOST_DEV
+  static TensorLayout GetFrameLayout(const TensorLayout &tl) {
+    return tl.skip('F');
+  }
+  DALI_HOST_DEV
+  static TensorLayout GetSequenceLayout(const TensorLayout &tl) {
+    if (tl.contains('F'))
+      return tl;
+    if (tl[0] == 'N')
+      return "NF" + tl.sub(1);
+    else
+      return "F" + tl;
+  }
+};
+
+/** @brief Calculates mapping of dimensions from given output to input.
+ *
+ * Array element corresponding to given output dimensions is the index of that
+ * dimension in the input layout.
+ *
```

Comment:
so calculates the result that for all valid i:
```
input_layout[result[i]] == output_layout[i]
```
I would name them the other way round, but whatever. Just add the formula, please.

Reply:
It was taken from live code in crop / slice.